swift/ql/lib/codeql/swift/controlflow/internal/Scope.qll

```diff
@@ -27,224 +28,13 @@ class Scope extends AstNode instanceof Scope::Range {
 
 cached
 private module Cached {
-  private AstNode getChild(AstNode ast) {
-    result = ast.(TopLevelCodeDecl).getBody()
-    or
-    result = ast.(AnyTryExpr).getSubExpr()
-    or
-    result = ast.(ClosureExpr).getBody()
-    or
-    result = ast.(ExplicitCastExpr).getSubExpr()
-    or
-    result = ast.(ForceValueExpr).getSubExpr()
-    or
-    result = ast.(IdentityExpr).getSubExpr()
-    or
-    result = ast.(ImplicitConversionExpr).getSubExpr()
-    or
-    result = ast.(InOutExpr).getSubExpr()
-    or
-    result = ast.(MemberRefExpr).getBaseExpr()
-    or
-    result = ast.(SelfApplyExpr).getBaseExpr()
-    or
-    result = ast.(VarargExpansionExpr).getSubExpr()
-    or
-    result = ast.(BindingPattern).getSubPattern()
-    or
-    result = ast.(EnumElementPattern).getSubPattern()
-    or
-    result = ast.(ExprPattern).getSubExpr()
-    or
-    result = ast.(IsPattern).getSubPattern()
-    or
-    result = ast.(OptionalSomePattern).getSubPattern()
-    or
-    result = ast.(ParenPattern).getSubPattern()
-    or
-    result = ast.(TypedPattern).getSubPattern()
-    or
-    result = ast.(DeferStmt).getBody()
-    or
-    result = ast.(DoStmt).getBody()
-    or
-    result = ast.(ReturnStmt).getResult()
-    or
-    result = ast.(ThrowStmt).getSubExpr()
+  private AstNode getParentOfAstStep(Element e) {
+    not e instanceof AstNode and
```

Comment:
this predicate here is because there may be edges in the parent/children tree traversing non-AstNodes? Which ones?

Reply:
anyway it might be good to comment why we need this :slightly_smiling_face:

Reply:
Oops, you're right. The return type should indeed be `Element`.

Reply:
> There's also ConditionElement I think. We could promote them both to AstNode though. It sounds kinda weird that the child of an AstNode is not an AstNode, doesn't it?

Indeed, it does. I think they should both be `AstNode`s, yeah.

Reply:
anyway, I'm ok with both leaving this as it is or expanding `AstNode` if you think it's the way to go.